src/State/Scope.php

```diff
@@ -0,0 +1,310 @@
+<?php
+
+/*
+ * This file is part of Raven.
+ *
+ * (c) Sentry Team
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sentry\State;
+
+use Sentry\Breadcrumbs\Breadcrumb;
+use Sentry\Context\Context;
+use Sentry\Context\TagsContext;
+use Sentry\Context\UserContext;
+use Sentry\Event;
+use Sentry\Severity;
+
+/**
+ * The scope holds data that should implicitly be sent with Sentry events. It
+ * can hold context data, extra parameters, level overrides, fingerprints etc.
+ */
+final class Scope
+{
+    /**
+     * @var Breadcrumb[] The list of breadcrumbs recorded in this scope
+     */
+    private $breadcrumbs = [];
+
+    /**
+     * @var UserContext The user data associated to this scope
+     */
+    private $user;
+
+    /**
+     * @var TagsContext The list of tags associated to this scope
+     */
+    private $tags;
+
+    /**
+     * @var Context A set of extra data associated to this scope
+     */
+    private $extra;
+
+    /**
+     * @var string[] List of fingerprints used to group events together in
+     *               Sentry
+     */
+    private $fingerprint = [];
+
+    /**
+     * @var Severity|null The severity to associate to the events captured in
+     *                    this scope
+     */
+    private $level;
+
+    /**
+     * @var callable[] List of event processors
+     */
+    private $eventProcessors = [];
+
+    /**
+     * Constructor.
+     */
+    public function __construct()
+    {
+        $this->user = new UserContext();
+        $this->tags = new TagsContext();
+        $this->extra = new Context();
+    }
+
+    /**
+     * Sets a new tag in the tags context.
+     *
+     * @param string $key   The key that uniquely identifies the tag
+     * @param string $value The value
+     *
+     * @return $this
+     */
+    public function setTag(string $key, string $value): self
+    {
+        $this->tags[$key] = $value;
+
+        return $this;
+    }
+
+    /**
+     * Gets the tags contained in the tags context.
+     *
+     * @return array<string, string>
+     */
+    public function getTags(): array
+    {
+        return $this->tags->toArray();
```

Comment:
Why all contexts are returned as array? To avoid editing from a distance?

Reply:
We're discussing about this here: https://github.com/getsentry/sentry-php/pull/679#discussion_r227275312